src/OpenTelemetry.Instrumentation.AWSLambda/Implementation/AWSMessagingUtils.cs

```diff
@@ -0,0 +1,137 @@
+            // https://docs.aws.amazon.com/sns/latest/dg/sns-large-payload-raw-message-delivery.html
+            SNSEvent.SNSMessage snsMessage = GetSnsMessage(sqsMessage);
+            if (snsMessage != null)
+            {
+                parentContext = ExtractParentContext(snsMessage);
+            }
+        }
+
+        return parentContext;
+    }
+
+    internal static PropagationContext ExtractParentContext(SNSEvent snsEvent)
+    {
+        // We assume there can be only one parent that's why we consider only a single (the last) record as the carrier.
+        var record = snsEvent?.Records?.LastOrDefault();
+        return ExtractParentContext(record);
+    }
+
+    internal static PropagationContext ExtractParentContext(SNSEvent.SNSRecord record)
+    {
+        return (record?.Sns?.MessageAttributes != null) ?
+            Propagators.DefaultTextMapPropagator.Extract(default, record.Sns.MessageAttributes, SnsMessageAttributeGetter) :
+            default;
+    }
+
+    internal static PropagationContext ExtractParentContext(SNSEvent.SNSMessage message)
+    {
+        return (message?.MessageAttributes != null) ?
+            Propagators.DefaultTextMapPropagator.Extract(default, message.MessageAttributes, SnsMessageAttributeGetter) :
+            default;
+    }
+
+    private static IEnumerable<string> SqsMessageAttributeGetter(IDictionary<string, SQSEvent.MessageAttribute> attributes, string attributeName)
+    {
+        if (!attributes.TryGetValue(attributeName, out var attribute))
+        {
+            return null;
+        }
+
+        return attribute?.StringValue != null ?
+            new[] { attribute.StringValue } :
+            attribute?.StringListValues;
+    }
+
+    private static IEnumerable<string> SnsMessageAttributeGetter(IDictionary<string, SNSEvent.MessageAttribute> attributes, string attributeName)
+    {
+        if (!attributes.TryGetValue(attributeName, out var attribute))
+        {
+            return null;
+        }
+
+        switch (attribute?.Type)
+        {
+            case SnsAttributeTypeString when attribute.Value != null:
+                return new[] { attribute.Value };
+            case SnsAttributeTypeStringArray when attribute.Value != null:
+                // Multiple values are stored as CSV (https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html).
+                return attribute.Value.Split(',');
+            default:
+                return null;
+        }
+    }
+
+    private static SNSEvent.SNSMessage GetSnsMessage(SQSEvent.SQSMessage sqsMessage)
+    {
+        SNSEvent.SNSMessage snsMessage = null;
+
+        var body = sqsMessage.Body;
+        if (body != null &&
+            body.TrimStart().StartsWith("{") &&
+            body.Contains(SnsMessageAttributes))
+        {
+            try
+            {
+                snsMessage = JsonConvert.DeserializeObject<SNSEvent.SNSMessage>(body);
```

Comment:
Is it possible to use System.Text.Json to do this as it would probably be lighter weight?

Reply:
I checked other projects and seems they use the Newtonsoft serializer (for example, see https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/e280873085ea96712cc636f00941c0e7acef7f3e/src/OpenTelemetry.Contrib.Extensions.AWSXRay/Resources/ResourceDetectorUtils.cs#L60). May be it's worth to keep Newtonsoft  for consistency and check/switch to the System.Text.Json everywhere. It may be a subject of a separate issue.